pallets/pallet-bonded-coins/src/tests/transactions/create_pool.rs

```diff
@@ -1 +1,251 @@
+use frame_support::{assert_err, assert_ok};
+use frame_system::{pallet_prelude::OriginFor, RawOrigin};
+use pallet_assets::{Error as AssetsPalletErrors, Event as AssetsPalletEvents};
+use sp_runtime::{ArithmeticError, BoundedVec};
+use sp_std::ops::Sub;
 
+use crate::{
+	mock::runtime::*, mock::*, types::PoolStatus, Event as BondingPalletEvents, NextAssetId, Pools, TokenMetaOf,
+};
+
+// create_pool tests
+
+#[test]
+fn single_currency() {
+	let initial_balance = 100_000_000_000_000_000u128;
+	ExtBuilder::default()
+		.with_native_balances(vec![(ACCOUNT_00, initial_balance)])
+		.with_collaterals(vec![DEFAULT_COLLATERAL_CURRENCY_ID])
+		.build()
+		.execute_with(|| {
+			assert_eq!(NextAssetId::<Test>::get(), 0);
+			assert_eq!(initial_balance, Balances::free_balance(ACCOUNT_00));
+
+			let origin = RawOrigin::Signed(ACCOUNT_00).into();
+			let curve = get_linear_bonding_curve_input();
+
+			let bonded_token = TokenMetaOf::<Test> {
+				name: BoundedVec::truncate_from(b"Bitcoin".to_vec()),
+				symbol: BoundedVec::truncate_from(b"btc".to_vec()),
+				min_balance: 1,
+			};
+
+			assert_ok!(BondingPallet::create_pool(
+				origin,
+				curve,
+				DEFAULT_COLLATERAL_CURRENCY_ID,
+				BoundedVec::truncate_from(vec![bonded_token]),
+				10,
+				true
+			));
+
+			let pool_id = calculate_pool_id(&[0]);
+
+			let details = Pools::<Test>::get(&pool_id).unwrap();
+
+			assert!(details.is_owner(&ACCOUNT_00));
+			assert!(details.is_manager(&ACCOUNT_00));
+			assert!(details.transferable);
+			assert_eq!(details.state, PoolStatus::Locked(Default::default()));
+			assert_eq!(details.denomination, 10);
+			assert_eq!(details.collateral_id, DEFAULT_COLLATERAL_CURRENCY_ID);
+			assert_eq!(details.bonded_currencies.len(), 1);
+			assert_eq!(details.bonded_currencies[0], 0);
+
+			assert_eq!(NextAssetId::<Test>::get(), 1);
+
+			assert_eq!(
+				Balances::free_balance(ACCOUNT_00),
+				initial_balance.sub(BondingPallet::calculate_pool_deposit(1))
+			);
+
+			System::assert_has_event(BondingPalletEvents::PoolCreated { id: pool_id.clone() }.into());
+
+			// TODO: check events or storage of linked pallets?
+			System::assert_has_event(
```

Comment:
So the `crate_pool` generates a `ForceCreated` event? Why is that?

Reply:
Turns out the assets pallet performs a 'force create' in the implementation of the `fungibles::create` trait method:
https://docs.rs/pallet-assets/40.0.0/src/pallet_assets/impl_fungibles.rs.html#187

Reply:
Ok but these are unit tests, you should not test storage and events generated by other pallets. Also because if you swap out the implementation of Fungibles with a different one, the events would be gone.

Reply:
Yeah this section kinda goes beyond the scope of unit tests, but because the pallet is 80% calls to other pallets, we need to at least test that these are called correctly - which seems quite hard to do in Substrate? We'd need to somehow _actually_ mock the assets pallet (right now we use its real implementation to test against) and be able to run assertions on the arguments with which trait methods are being called, which I'm not sure we can do? So we need to make a call on how we want to test here - either mock away the fungibles trait implementation and capture calls, or expand the scope of the tests to be more integration-testy.

Reply:
I assume we have no equivalent to jests `spyOn` or auto-mock functionality here?

Reply:
No, that would have to be implemented in the trait mock implementation, where you keep track yourself. In general, I am not a very big fan of using existing stuff for unit tests, but rather mock everything I need to mock. In this case I think it's fine to use the assets pallet, but I would not check for the events generates, and rather use the same trait bounds you define in the config trait to verify that the conditions are met. Meaning, calling asserts only on functions that are exposed via the traits that the config object is required to implement. So checking the balance of a user, for instance, would not call the assets pallet directly, but would get the value via the trait that is responsible for returning that value. That would be a good compromise, I think.

Reply:
makes sense! Let me see if I can figure out how to call into the asset pallet's fungibles trait from the tests